src/console/man/om_project.1
.\" Text automatically generated by txt2man
.TH om_project  "04 September 2013" "" ""
.SH NAME
\fBom_project \fP- project a distribution model using the openModeller framework
\fB
.SH SYNOPSIS
.nf
.fam C
     \fBom_project\fP [-] \fIv\fP \fB--version\fP | [ \fIr\fP \fB--xml-req\fP \fIXML_REQUEST_FILE\fP | \fIo\fP \fB--model\fP \fIFILE\fP [ \fIt\fP \fB--template\fP \fIFILE\fP ] [ \fIf\fP \fB--format\fP \fIOUTPUT_TYPE\fP ] ] \fIm\fP \fB--dist-map\fP \fIFILE\fP [ \fB--log-level\fP \fILEVEL\fP ] [ \fB--log-file\fP \fIFILE\fP ] [ \fB--prog-file\fP \fIFILE\fP ] [ \fB--stat-file\fP \fIFILE\fP ]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBom_project\fP is a command line tool to project distribution models. The main input can be an XML file containing a projection request according to the ProjectionParameters element definition in http://openmodeller.cria.org.br/xml/1.0/openModeller.xsd (see also projection_request.xml in the openModeller examples directory). The second option can only be used in native projections, i.e., when the layers used in the projection are the same as the layers used in model creation. In this case it is possible to specify a serialized model file with an optional template file and an optional file format. Projection layers will be taken from the model creation layers stored in the serialized model. When no template file is specified, the first layer is taken as a template. Template layers determine the cell size and the spatial reference that will be used by the distribution map. Valid values for file format are:
.PP
GreyTiff = grey scale GeoTiff (0 <= cell value <= 255). (*.tif)
GreyTiff100 = grey scale GeoTiff (0 <= cell value <= 199). (*.tif)
FloatingTiff = floating point GeoTiff (cell value = probability of presence) (*.tif)
GreyBMP = grey scale BMP. (*.bmp)
FloatingHFA = Erdas Imagine Floating Point (cell value = probability of presence)
NoData will be written as \fB-1\fP. (*.img) 
ByteHFA = Erdas Imagine Byte representation (0 <= cell value <= 100).
NoData will be written as 101. Default format. (*.img) 
.PP
The distribution map (raster file) generated by openModeller will be stored in another file specified as a parameter. To generate a distribution model, which is part of the projection parameters, use om_model.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIr\fP, \fB--xml-req\fP
File containing a projection request in XML.
.TP
.B
-\fIo\fP, \fB--model\fP
Serialized model (native projection).
.TP
.B
-\fIt\fP, \fB--template\fP
Template raster file (native projection).
.TP
.B
-\fIf\fP, \fB--format\fP
Distribution map file format (native projection).
.TP
.B
-\fIm\fP, \fB--dist-map\fP
File where the generated raster will be stored.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.TP
.B
\fB--log-file\fP
File to store log information.
.TP
.B
\fB--prog-file\fP
File to store progress (\fB-1\fP=queued, \fB-2\fP=aborted, \fB-3\fP=cancelled, [0,100]=progress).
.TP
.B
\fB--stat-file\fP
File to store projection statistics (in XML).
.PP
\fB-c\fP, \fB--config-file\fP Configuration file for openModeller (available since version 1.4).
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.4.0 (?-?-?)
-------------
SVN revision: ?

	* New CONSENSUS algorithm that can be used to generate models with different algorithms by converting each individual model into a binary model given a threshold and then "merging" them into a final combined model according to a specified level of agreement.
	* Upgraded gSOAP version being used by OMWS to 2.8.13.
	* Included library API versioning, starting with 0 from this release.
	* Moved some of the compiler definitions from CMake to internal #defines, so that client programs can include openModeller headers without worrying about adding specific definitions if they want to compile directly without using CMake.
	* New command-line tool om_evaluate to return raw model values given a set of points and an environmental scenario.
	* Fixed segfault when needNormalization was called on a deserialized SVM algorithm.
	* Modified om_pseudo command-line tool to also accept XML input (in this case also producing XML output), optionally storing it into a file. When a model is specified as a parameter, presences are generated inside the model, while absences are generated outside.

Release 1.3.0 (2012-07-23)
-------------
SVN revision: 5592

	* New version of OM-MAXENT compatible with Maxent 3.3.3e.
	* Changed the deserialization code to accept XML with namespace prefixes (they are now ignored if present).
	* Fixed bug in ANN algorithm which was limiting the number of input points to 10k (seg faulting otherwise).
	* New documentation about the web service interface (OMWS), aka SOAP interface.
	* New documentation about oM Server (server-side implementation of OMWS).
	* Fixed bug in oM Server getProgress operation which was sometimes returning 100% when the job was still unfinished.
	* New oM Server configuration/script allowing automatic synchronization of layers.
	* Several improvements in the OMWS Perl command-line client.
	* Several adjustments in the SWIG Python interface, exposing more methods and objects.
	* Fixed memory leak in the Environmental Distance algorithm when the Mahalanobis metric was used after model deserialization.
	* New WCS proxy raster driver that allows WCS rasters to be identified through strings in the format: wcs>wcs_service_url>coverage_id

Release 1.2.0 (2011-07-21)
--------------------------
SVN revision: 5330

	* Changed the ENFA algorithm to force marginality to be the first factor.  Changed broken stick distribution to work on specialisation factors only.  Add new parameter to enable log display of variable loadings for retained factors.
	* om_console now accepts a second parameter (after the request file) to specify the log level.
	* Changed the ENFA algorithm allowing absence data to be treated as background points, and log transforming the geomean calculation to avoid the occasional float overflow with large numbers of presences. Fixed a bug in the RetainVariation parameter (range changed from 50-100 to 0.5-1.0). Also fixed a deserialization error.
	* Compilation now requires CMake 2.6.4.
	* New algorithm Random Forests.
	* Web service interface now accepts a new optional parameter specifying the prediction threshold to be used when calculating area statistics during projection.
	* Renamed envelope element in the getProjectionMetadata response of the web service API to distinguish it from the subelement.
	* Added control to a maximum number of concurrent jobs on the modelling server.
	* Environmental values can now be loaded from the Sampler XML when provided.
	* Web service interface now accepts optional parameters: IgnoreAbsences for the confusion matrix and UseAbsencesAsBackground for the ROC curve calculations.
	* Added support to gzip compression in the Web service interface.
	* Changed the Environmental Distance algorithm to return chi-square distribution probabilities when the metric is Mahalanobis and the maximum distance is 1.
	* Changed the Maxent algorithm allowing absence data to be treated as background points. Also fixed memory deallocation issues.

Release 1.1.0 (2010-02-05)
--------------------------
SVN revision: 5114

	* Bugfix: Interquartile and envelope size adjustments were using wrong reference values for salinity and temperature in the Aquamaps algorithm.
	* Bugfix: All GARP algorithms were crashing when the abundance value associated with one of the occurrence points had more than two digits in the integer part.
	* Increased the default number of neurons in the hidden layer of the ANN algorithm.
	* New argument "s" in om_niche allowing model outputs to be scaled.
	* Re-implemented and re-enabled Maximum entropy algorithm based on the well-known Maxent.
	* Changed environmental distance algorithm forcing the Mahalanobis distance to use an identity matrix as the covariance matrix when there's only one input point.
	* Fixed small logical error in the CSM algorithm that could affect models with a large number of layers.
	* New algorithm (Niche Mosaic) implemented by Dr. Missae Yamamoto.
	* New algorithm (ENFA) kindly implemented by Dr. Chris Yesson.

Release 1.0 (2009-05-22)
------------------------
SVN revision: 5035

	* Changed Artificial Neural Networks algorithm to avoid exceeding a maximum number of iterations when training by minimum error (parameter Epoch is used to specify the maximum number of iterations).
	* Added new method in the ConfusionMatrix class to use the lowest probability among the training presences as the threshold.
	* Changed the modelling protocol to allow createModel and testModel requests to specify what statistics should be calculated.
	* Removed model statistics parameters from om_model since they can be specified as part of the XML request.
	* All command line programs now return the same openModeller version number.
	* Changed om_niche parameters (now using getopts) and created a man page for it.
	* Implemented custom log file in the modelling server.

src/console/om_project.cpp
/*************************************************************************************
 * This code simply loads a previously created model, loads a reprojection 
 * environment, and projects the model into that environment,
 * then writes the model out.
 *    -------------------
 *       begin                : November 2005
 *       copyright            : (C) 2005 by T.Sutton, Kevin Ruland, Renato De Giovanni
 *       email                : tim@linfiniti.com
 *************************************************************************************/

/***************************************************************************
 *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <openmodeller/om.hh>
#include <openmodeller/Log.hh>
#include <openmodeller/os_specific.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <fstream>   // file I/O for XML
#include <sstream>   // ostringstream datatype
#include <stdio.h>   // file I/O for log
#include <time.h>    // used to limit the number of times that the progress is written to a file
#include <string>    // string library
#include <stdexcept> // try/catch

#ifdef MPI_FOUND
#include "mpi.h"
#endif

using namespace std;

int main( int argc, char **argv ) {

  Options opts;
  int option;

  // command-line parameters (short name, long name, description, take args)
  opts.addOption( "v", "version"    , "Display version info"                        , false );
  opts.addOption( "r", "xml-req"    , "Projection request file in XML"              , true );
  opts.addOption( "o", "model"      , "File with serialized model (native projection)", true );
  opts.addOption( "t", "template"   , "Raster template for the distribution map (native projection)", true );
  opts.addOption( "f", "format"     , "File format for the distribution map (native projection)", true );
  opts.addOption( "m", "dist-map"   , "File to store the generated model"           , true );
  opts.addOption( "" , "log-level"  , "Set the log level (debug, warn, info, error)", true );
  opts.addOption( "" , "log-file"   , "Log file"                                    , true );
  opts.addOption( "" , "prog-file"  , "File to store projection progress"           , true );
  opts.addOption( "" , "stat-file"  , "File to store projection statistics"         , true );
  opts.addOption( "c", "config-file", "Configuration file for openModeller"         , true );

  std::string log_level("info");
  std::string request_file;
  std::string model_file;
  std::string tmpl_file;
  std::string format;
  std::string map_file;
  std::string log_file;
  std::string progress_file;
  std::string statistics_file;
  std::string config_file;

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] ); 
    exit(0);
  }

  // Set up any related external resources
  setupExternalResources();

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        printf( "om_project %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 1:
        request_file = opts.getArgs( option );
        break;
      case 2:
        model_file = opts.getArgs( option );
        break;
      case 3:
        tmpl_file = opts.getArgs( option );
        break;
      case 4:
        format = opts.getArgs( option );
        break;
      case 5:
        map_file = opts.getArgs( option );
        break;
      case 6:
        log_level = opts.getArgs( option );
        break;
      case 7:
        log_file = opts.getArgs( option );
        break;
      case 8:
        progress_file = opts.getArgs( option );
        break;
      case 9:
        statistics_file = opts.getArgs( option );
        break;
      case 10:
        config_file = opts.getArgs( option );
        break;
      default:
        break;
    }
  }

  // om configuration
  if ( ! config_file.empty() ) { 

    Settings::loadConfig( config_file );
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  if ( ! log_file.empty() ) {

    Log::instance()->set( level_code, log_file, "" );
  }
  else {
 
    // Just set the level - things will go to stderr
    Log::instance()->setLevel( level_code );
  }

  // Check parameters

  if ( request_file.empty() && model_file.empty() ) {

    printf( "Please specify either a projection request file in XML or a serialized model\n");
    exit(-1);
  }

  if ( (! request_file.empty()) && ( (! model_file.empty()) || (! tmpl_file.empty()) || (! format.empty() ) ) ) {

    Log::instance()->warn( "When a request file is specified, --model, --template and --format are ignored" );
  }

  if ( map_file.empty() ) {

    printf( "Please specify a file to store the distribution map\n");
    exit(-1);
  }

  // Initialize progress data if user wants to track progress
  progress_data prog_data;

  if ( ! progress_file.empty() ) { 

    prog_data.file_name = progress_file;

    time( &prog_data.timestamp );

    prog_data.progress = -1.0; // queued

    // Always create initial file with progress 0
    progressFileCallback( 0.0, &prog_data );
  }

  // Real work

  try {

    #ifdef MPI_FOUND
      Log::instance()->info( "Running parallel projection version\n" );
      MPI_Init( &argc, &argv );
      int rank;
      MPI_Comm_rank( MPI_COMM_WORLD, &rank );
      Log::instance()->debug( "Rank of calling process: %d\n", rank );
      MPI_Barrier( MPI_COMM_WORLD );
    #else
      Log::instance()->info( "Running serial projection version\n" );
    #endif

    // Load algorithms and instantiate controller class
    AlgorithmFactory::searchDefaultDirs();

    // If user wants to track progress
    if ( ! progress_file.empty() ) { 

      // Set callback to write to a file
      om.setMapCallback( progressFileCallback, &prog_data );
    }
    else if ( ! statistics_file.empty() ) {

      // Default callback will display progress on screen when a statistics file was specified
      // (which means statistics won't be sent to stdout)
      om.setMapCallback( progressDisplayCallback );
    }

    std::ostringstream model_output;

    if ( ! request_file.empty() ) {

      ConfigurationPtr input = Configuration::readXml( request_file.c_str() );
      om.setProjectionConfiguration( input );
      om.createMap( map_file.c_str() );
    }
    else {

      // Native projection - get original environment from serialized model
      ConfigurationPtr input = Configuration::readXml( model_file.c_str() );

      om.setModelConfiguration( input );

      EnvironmentPtr env = om.getEnvironment();

      MapFormat tmpl;

      if ( tmpl_file.empty() ) {

        // Use first layer as reference
        std::string first_layer = env->getLayerPath(0);

        tmpl = MapFormat( first_layer.c_str() );
      }
      else {

        tmpl = MapFormat( tmpl_file.c_str() );
      }

      if ( ! format.empty() ) {

        tmpl.setFormat( format );
      }

      om.createMap( env, map_file.c_str(), tmpl );
    }

    AreaStats * stats = om.getActualAreaStats();

    ConfigurationPtr stats_cfg = stats->getConfiguration();

    std::ostringstream statistics_output;

    Configuration::writeXml( stats_cfg, statistics_output );

    std::cerr << flush;

    // Write statistics output to file, if requested
    if ( ! statistics_file.empty() ) {

      ofstream file( statistics_file.c_str() );
      file << statistics_output.str();
      file.close();
    }
    else {

      // Otherwise send it to stdout
      std::cout << statistics_output.str().c_str() << endl << flush;
    }

    delete stats;

    // If user wants to track progress
    if ( ! progress_file.empty() ) { 

      // Check if job was completed
      if ( prog_data.progress != 1 ) {

        // -2 means aborted
        progressFileCallback( -2.0, &prog_data );
      }
    }
  }
  catch ( runtime_error e ) {

    // If user is tracking progress
    if ( ! progress_file.empty() ) { 

      // -2 means aborted
      progressFileCallback( -2.0, &prog_data );
    }

    printf( "om_project aborted: %s\n", e.what() );
  }

  #ifdef MPI_FOUND
  MPI_Barrier( MPI_COMM_WORLD ); 
  MPI_Finalize();
  #endif

}
